contracts/ColonyNetworkStaking.sol

```diff
@@ -0,0 +1,306 @@
+    // Move some over.
+    clny.transferFrom(msg.sender, this, _amount);
+    // Check it actually transferred
+    assert(clny.balanceOf(this)-networkBalance==_amount);
+    // Note who it belongs to.
+    stakedBalances[msg.sender] += _amount;
+  }
+
+  function withdraw(uint _amount) public {
+    uint256 balance = stakedBalances[msg.sender];
+    require(balance >= _amount);
+    bytes32 submittedHash;
+    (submittedHash, ) = ReputationMiningCycle(reputationMiningCycle).hasSubmitted(msg.sender);
+    bool hasRequesterSubmitted = submittedHash == 0x0 ? false : true;
+    require(hasRequesterSubmitted==false);
+    stakedBalances[msg.sender] -= _amount;
+    Token clny = Token(IColony(_colonies["Common Colony"]).getToken());
+    clny.transfer(msg.sender, _amount);
+  }
+
+  function getStakedBalance(address _user) public view returns (uint) {
+    return stakedBalances[_user];
+  }
+
+  function setReputationRootHash(bytes32 newHash, uint256 newNNodes, address[] stakers) public {
+    require(msg.sender == reputationMiningCycle);
+    reputationRootHash = newHash;
+    reputationRootHashNNodes = newNNodes;
+    // Clear out the inactive reputation log. We're setting a new root hash, so we're done with it.
+    delete ReputationUpdateLog[(activeReputationUpdateLog + 1) % 2];
+    // The active reputation update log is now switched to be the one we've just cleared out.
+    // The old activeReputationUpdateLog will be used for the next reputation mining cycle
+    activeReputationUpdateLog = (activeReputationUpdateLog + 1) % 2;
+    // Reward stakers
+    rewardStakers(stakers);
+    reputationMiningCycle = 0x0;
+    startNextCycle();
+  }
+
+  function startNextCycle() public {
+    require(reputationMiningCycle == 0x0);
+    reputationMiningCycle = new ReputationMiningCycle();
+  }
+
+  function getReputationMiningCycle() public view returns(address) {
+    return reputationMiningCycle;
+  }
+
+  function punishStakers(address[] stakers) public {
+    // TODO: Actually think about this function
+    // Passing an array so that we don't incur the EtherRouter overhead for each staker if we looped over
+    // it in ReputationMiningCycle.invalidateHash;
+    require(msg.sender == reputationMiningCycle);
+    for (uint256 i = 0; i < stakers.length; i++) {
+      // This is pretty harsh! Are we happy with this?
+      // Alternative: lose more than they would have gained for backing the right hash.
+      stakedBalances[stakers[i]] = 0;
+    }
+    // TODO: Where do these staked tokens go? Maybe split between the person who did the 'invalidate' transaction
+    // and the colony network?
+    // TODO: Lose rep?
+  }
+
+  function rewardStakers(address[] stakers) internal {
+    // Internal unlike punish, because it's only ever called from setReputationRootHash
+
+    // TODO: Actually think about this function
+    // Passing an array so that we don't incur the EtherRouter overhead for each staker if we looped over
```

Comment:
Copy and paste leftover to remove

Reply:
Though maybe not as much removed as you're expecting.